ipam/client/backends/phpipam.py

```diff
@@ -131,6 +144,91 @@ def get_allocated_ips_by_subnet_id(self, subnetid):
+        parent_subnet_used_ips = self.get_allocated_ips_by_subnet_id(
+            parent_subnet_id)
+
+        if len(parent_subnet_used_ips):
+            raise ValueError('Parent subnet {} must not contain any '
+                             'allocated IP address!'.format(parent_subnet))
+
+        self.cur.execute(
+            'INSERT INTO subnets '
+            '(subnet, mask, sectionId, description, vrfId, '
+            'masterSubnetId, vlanId, permissions) '
+            'VALUES (\'{:d}\', \'{}\', \'{}\', \'{}\', '
+            '\'{}\', \'{}\', \'{}\', \'{}\')'.format(
+                int(subnet.ip),
+                subnet.prefixlen,
+                self.section_id,
+                description,
+                self.subnet_options['vrf_id'],
+                parent_subnet_id,
+                self.subnet_options['vlan_id'],
+                self.subnet_options['permissions']))
+        return subnet
+
+    def _get_next_free_subnet(self, subnet, prefixlen):
+        """
+        Find next free prefixlen-wide subnet in given subnet.
+        """
+        subnet_id = self.find_subnet_id(subnet)
+        if not subnet_id:
+            raise ValueError('Unable to get subnet id from database '
+                             'for subnet {}'.format(subnet))
+
+        unavailable_subnets = self._get_unavailable_subnets(
+            subnet_id, prefixlen)
+
+        for candidate_subnet in subnet.subnet(prefixlen):
+            if candidate_subnet not in unavailable_subnets:
+                return candidate_subnet
+
+        return None
+
+    def _get_unavailable_subnets(self, subnet_id, prefixlen):
+        """
+        Return list of allocated subnets in a parent subnet, given its id
+        """
+        self.cur.execute('SELECT subnet, mask FROM subnets '
+                         'WHERE masterSubnetId={} '
+                         'ORDER BY subnet ASC'.format(subnet_id))
+
+        unavailable_subnets = []
+        for row in self.cur:
+            used_subnet_network = IPAddress(int(row[0]))
+            used_subnet = IPNetwork('{}/{}'.format(
+                used_subnet_network, int(row[1])))
+            # If any portion of a prefixlen-wide subnet is
+            # allocated, then it is not available
+
+            # allocated subnet is smaller than wanted prefixlen, consider
+            # prefixlen-wide subnet as unavailable as it might be in the
+            # middle of the target subnet
+            if used_subnet.prefixlen >= prefixlen:
+                network = IPNetwork('{}/{}'.format(
+                    used_subnet.network, prefixlen)).network
+                unavailable_subnets.append(IPNetwork('{}/{}'.format(
+                    network, prefixlen)))
+            # allocated subnet is wider than prefixlen, split it into
+            # multiple prefixlen subnets
+            else:
+                used_subnets = used_subnet.subnet(prefixlen)
+                unavailable_subnets.extend(used_subnets)
+
```

Comment:
Same comment as above. Such tooling would avoid doing these tricks to split address space into multiple subnets of prefixlen size.

Reply:
Fixed by switching to ipaddress which provides an overlaps method.